File: problems/halim/subsets.py
from typing import List
import os


def subset(n) -> List[List[int]]:
    nums = [i for i in range(1, n+1)]
    res = []
    path = []

    def dfs(i):
        if i >= len(nums):
            res.append(path.copy())
            return

        path.append(nums[i])
        dfs(i+1)

        path.pop()
        dfs(i+1)

    dfs(0)
    return res


def main():
    n = 20
    res = subset(n)
    script_dir = os.path.dirname(os.path.abspath(__file__))
    output_path = os.path.join(script_dir, "subsets.txt")

    with open(output_path, "w") as f:
        for r in res:
            f.write(str(r) + "\n")


if __name__ == "__main__":
    main()

File: problems/halim/subsets.cpp
#include "header.h"
using namespace std;

vector<vector<int>> subsets(int n) {
    vector<int> nums;
    for (int i = 1; i <= n; i++) {
        nums.push_back(i);
    }
    vector<int> path;
    vector<vector<int>> res;

    function<void(int)> dfs = [&](int i) {
        if (i >= (int)nums.size()) {
            res.push_back(path);
            return;
        }
        path.push_back(nums[i]);
        dfs(i + 1);
        path.pop_back();
        dfs(i + 1);
    };
    dfs(0);
    return res;
}

int main() {
    int n = 20;
    vector<vector<int>> res = subsets(n);
    ofstream fout("subsets.txt");
    for (auto &p : res) {
        for (int x : p) {
            fout << x << " ";
        }
        fout << "\n";
    }
    fout.close();
    return 0;
}